File: algorithm/src/main/java/baekjoon/p04592/Main.java
package baekjoon.p04592;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        while (true) {
            int n = sc.nextInt();
            if (n == 0) {
                break;
            }

            int[] inputs = new int[n];
            for (int i = 0; i < inputs.length; i++) {
                inputs[i] = sc.nextInt();
            }

            for (int solve : solve(inputs)) {
                System.out.print(solve + " ");
            }
            System.out.println("$");
        }
    }

    private static List<Integer> solve(int[] inputs) {
        List<Integer> ret = new ArrayList<>();
        int selected = -1;

        for (int input : inputs) {
            if (selected != input) {
                selected = input;
                ret.add(input);
            }
        }

        return ret;
    }

}

File: algorithm/src/main/java/baekjoon/p04592/Main.kt
package baekjoon.p04592

fun main(args: Array<String>) {
    while (true) {
        var inputs = readLine()!!.split(" ").map { it.toInt() }
        if (inputs[0] == 0) {
            break
        }

        inputs = inputs.filterIndexed { index, i -> index > 0 }
        println(solve(inputs).joinToString(" ").plus(" $"))
    }
}

private fun solve(inputs: List<Int>): List<Int> {
    val ret = ArrayList<Int>()
    var selected = -1

    for (input in inputs) {
        if (input != selected) {
            selected = input
            ret.add(input)
        }
    }

    return ret
}